net/sched/p4tc/p4tc_table.c

```diff
@@ -603,26 +709,177 @@ tcf_table_find_byany(struct p4tc_pipeline *pipeline, struct nlattr *name_attr,
 	return ERR_PTR(err);
 }
 
+static int
+tcf_table_init_default_act(struct net *net, struct nlattr **tb,
+			   struct p4tc_table_default_action **default_act,
+			   __u16 curr_permissions, struct netlink_ext_ack *extack)
+{
+	int ret;
+
+	if (!tb[P4TC_TABLE_DEFAULT_ACTION] && !tb[P4TC_TABLE_DEFAULT_PERMISSIONS])
+		return 0;
+
+	*default_act = kzalloc(sizeof(**default_act), GFP_KERNEL);
+	if (!(*default_act))
+		return -ENOMEM;
+
+	if (tb[P4TC_TABLE_DEFAULT_ACTION]) {
+		struct tc_action **default_acts;
+
+		if (!(curr_permissions & P4TC_PERMISSIONS_CONTROL_U)) {
+			NL_SET_ERR_MSG(extack,
+				       "Permission denied: Unable to update default hit action");
+			ret = -EPERM;
+			goto default_act_free;
+		}
+
+		default_acts = kcalloc(TCA_ACT_MAX_PRIO,
+				       sizeof(struct tc_action *),
+				       GFP_KERNEL);
+		if (!default_acts) {
+			ret = -ENOMEM;
+			goto default_act_free;
+		}
+
+		ret = p4tc_action_init(net, tb[P4TC_TABLE_DEFAULT_ACTION],
+				       default_acts, 0, extack);
+		(*default_act)->default_acts = default_acts;
+		if (ret < 0) {
+			kfree(default_acts);
+			goto default_act_free;
+		} else if (ret > 1) {
+			NL_SET_ERR_MSG(extack, "Can only have one hit action");
+			ret = -EINVAL;
+			goto default_acts_free;
+		}
+	}
+
+	if (tb[P4TC_TABLE_DEFAULT_PERMISSIONS]) {
+		__u16 *permissions;
+
+		permissions = nla_data(tb[P4TC_TABLE_DEFAULT_PERMISSIONS]);
+		if (*permissions > (1 << 10) - 1) {
+			NL_SET_ERR_MSG(extack,
+				       "Permission may only have 10 bits turned on");
+			ret = -EINVAL;
+			goto default_acts_free;
+		}
+		if (!(*permissions & P4TC_PERMISSIONS_DATA_X)) {
+			NL_SET_ERR_MSG(extack,
+				       "Default action must have data path execute permissions");
+			ret = -EINVAL;
+			goto default_acts_free;
+		}
+		(*default_act)->permissions = *permissions;
+	} else {
+		(*default_act)->permissions = curr_permissions;
+	}
+
+	return 0;
+
+default_acts_free:
+	if (*default_act && (*default_act)->default_acts) {
+		tcf_action_destroy((*default_act)->default_acts,
+				   TCA_ACT_UNBIND);
+		kfree((*default_act)->default_acts);
+	}
+
+default_act_free:
+	kfree(*default_act);
+
+	return ret;
+}
+
+static int
+tcf_table_init_default_acts(struct net *net, struct nlattr **tb,
+			    struct p4tc_table *table,
+			    struct p4tc_table_default_action **default_hitact,
+			    struct p4tc_table_default_action **default_missact,
+			    struct netlink_ext_ack *extack)
+{
+	struct nlattr *tb_default[P4TC_TABLE_DEFAULT_MAX + 1];
+	__u16 permissions = P4TC_PERMISSIONS_CONTROL | P4TC_PERMISSIONS_DATA;
+	int ret;
+
+	*default_missact = NULL;
+	*default_hitact = NULL;
```

Comment:
Why not allocate and then pass the allocated pointers?
Looks like it could simplify some things

Reply:
If you do that you will be allocating even when you don't have P4TC_TABLE_DEFAULT_HIT or P4TC_TABLE_DEFAULT_MISS attributes
I tried to at least concentrate all of the operations in one function (tcf_table_init_default_act), but probably that didn't simplify it enough
Do you have another suggestion?

Reply:
Now I'm wondering if it wouldn't be better to pass in table and a bool to differentiate create/update rather then using table as the bool and passing the default action fields regardless... Looks like we can cut down one parameter this way

Reply:
There are some cases where you don't want to write them directly to the table, such as update.
From what I understood, that's what you are suggesting
Did I understand correctly?

Reply:
I forgot this table is being read at datapath